src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
The only way to get this to work is:

1) Have the UI thread pump.  I AM LOATHE TO DO THIS.  To the point that someone would have to explain how it solved world peace before i even considered it.
2) Have dedicated points to do UI operations.  The problem with this is: Do you want to do the UI operations before?  After?  Both?  

We have some precedence here with CodeActionWithOptions.  It has a method we guarantee will get called on the UI thread ("GetOptions") before we call into GetOperationsAsync.

I think such a pattern would be acceptable in CodeFixProvider and CodeRefactoringProvider as well.